drafts/bitcartel-nu0-tx-format/draft.rst

```diff
@@ -0,0 +1,202 @@
+    // https://github.com/zcash/zcash/blob/30d3d2dfd438a20167ddbe5ed2027d465cbec2f0/src/gtest/test_checktransaction.cpp#L99
+
+By using a negative value for the version field, we ensure there is replay protection between Legacy and Overwinter compatible software.
+
+Consider an example where the raw little-endian bytes of the version field of an Overwinter transaction begin 0xFDFFFFFF ( binary 11111101111111111111111111111111 )
+
+Legacy parsers will deserialize the raw version field as a 32-bit signed integer.  With a negative version value of -3, legacy parsers will reject the transaction.
+
+Overwinter parsers will deserialize the raw version field as a 32-bit signed integer.  With a negative raw value of -3, the Overwinter parser will accept the transaction as the most significan bit of the 32-bit signed integer has been set.
+
+Overwinter parsers can retrieve the transaction format version of 3 by getting the absolute value of the raw version field e.g. using standard library call std::abs()
+
+Currently, the nVersion field is a public member variable which can be accessed directly.  As part of implementing Overwinter, the nVersion field will be made private with access restricted to using getters, e.g.::
+
+    bool isLegacyFormat()         // return true if the msb of nVersion is not set
+    unsigned int32 getVersion()   // return absolute value of raw version field which is compatible with Legacy and Overwinter
+
+
+Forwards Compatibility
+----------------------
+
+A branch may support many transaction version formats.  For example:
+
+* Zcash reference implementation, branch "Zcash", versions 3, 4.
+* Fork of Zcash, branch "Clone", versions 3, 4*
+
+Where transaction format version 4* for the "Clone" branch might be substantially different from the expected transaction format version 4 for the "Zcash" branch.
+
+Given forwards compatibility, we want the "Zcash" branch nodes to accept transaction version 4, whilst rejecting version 4* transactions which are intended only for the "Clone" branch.
+
+To achieve this, Overwinter requires a transaction to include a branch ID, to explicitly state which branch of the network this transaction is intended for.
+
+Overwinter introduces a new signature hashing scheme which includes the branch ID, but by including the branch ID into the transaction format, clients can quickly reject transactions during deserialization without having to check signatures.
+
+A simple way to filter transactions might look like this::
+
+    if (tx.branchID != CLIENT_BRANCH_ID) { ... }
+    
+However given that a branch may support a set of transaction version formats, we should implement such that we can write code like::
+
+    if (isBranchSupported(tx.getBranchID())) { ... }
+
+    if (tx.isSupportedBranch()) { ... }
+    
+Overwinter will introduce a method for developers to easily specify and update a map of supported branch IDs and transaction versions which can be easily accessed throughout the system.
```

Comment:
Per above, there will only ever be a single branch ID corresponding to a particular `nVersion` - specifically, the branch ID of the upgrade in which that `nVersion` was introduced.

Reply:
So with:
```
overwinter branch id = 100
sapling branch id = 200
```
The map of tx versions a branch supports might look like this
```
100 -> { 3 },
200 -> { 3 , 4 }
```
But it sounds like you want a map of version to branch id introduction.
```
3 -> 100,
4 -> 200
```
But this doesn't tell me if branch id 200 supports tx 3 or not, nor if branch id 100 supports tx 4 ?

Reply:
Whether branch id 200 supports 3 or not is a consensus rule. It is irrelevant to the local node being able to *parse* 3.

What I am suggesting is a map of version to branch ID introduction like so:
```
{ 3 } -> 100,
{ 4, 5 } -> 200
```

It's entirely possible for future branch ID 300 to make tx 3 and tx 4 invalid under this model.

Reply:
Rename as parser_branch_id ?

Reply:
Or ``format_branch_id``. Then the other usage for signature hash checking would be ``consensus_branch_id``.